Stratis.Bitcoin.Features.BlockStore/LoopSteps/ProcessPendingStorageStep.cs

```diff
@@ -43,132 +42,158 @@ internal ProcessPendingStorageStep(BlockStoreLoop blockStoreLoop, ILoggerFactory
+            this.logger.LogTrace("{0}:'{1}/{2}',{3}:{4}", nameof(nextChainedBlock), nextChainedBlock?.HashBlock, nextChainedBlock?.Height, nameof(disposeMode), disposeMode);
 
-            var context = new ProcessPendingStorageContext(this.BlockStoreLoop, nextChainedBlock);
+            var context = new ProcessPendingStorageContext(this.BlockStoreLoop, nextChainedBlock, cancellationToken);
 
             // Next block does not exist in pending storage, continue onto the download blocks step.
             if (!this.BlockStoreLoop.PendingStorage.ContainsKey(context.NextChainedBlock.HashBlock))
                 return StepResult.Next;
 
-            // If pending storage has not yet reached the threshold or the node is in IDB, stop the loop and wait for more blocks.
-            if (this.BlockStoreLoop.ChainState.IsInitialBlockDownload == true &&
-                disposeMode == false &&
-                this.BlockStoreLoop.PendingStorage.Count < BlockStoreLoop.PendingStorageBatchThreshold)
-                return StepResult.Stop;
+            if (disposeMode)
+                return await ProcessWhenDisposing(context);
+
+            if (this.BlockStoreLoop.ChainState.IsInitialBlockDownload)
+                return await ProcessWhenInIBD(context);
+
+            if (!this.BlockStoreLoop.ChainState.IsInitialBlockDownload)
+                return await ProcessWhenNotInIBD(context);
+
+            return StepResult.Stop;
+        }
+
+        /// <summary>
+        /// When the node disposes, process all the blocks in <see cref="BlockStoreLoop.PendingStorage"/> until 
+        /// its empty
+        /// </summary>
+        private async Task<StepResult> ProcessWhenDisposing(ProcessPendingStorageContext context)
+        {
+            this.logger.LogDebug("{0}:'{1}/{2}'", nameof(context.NextChainedBlock), context.NextChainedBlock.HashBlock, context.NextChainedBlock.Height);
 
-            while (!cancellationToken.IsCancellationRequested)
+            while (this.BlockStoreLoop.PendingStorage.Count > 0)
             {
-                var blockIsInPendingStorage = this.BlockStoreLoop.PendingStorage.TryRemove(context.NextChainedBlock.HashBlock, out context.PendingBlockPairToStore);
-                if (blockIsInPendingStorage == false)
-                {
-                    if (context.PendingBlockPairsToStore.Count == 0)
-                        return StepResult.Next;
+                PrepareNextBlockFromPendingStorage(context);
 
-                    if (context.PendingBlockPairsToStore.Any())
-                    {
-                        this.logger.LogTrace("[FLUSH_NEXTBLOCKNOTINPENDING]", nameof(context.PendingStorageBatchSize), context.PendingStorageBatchSize);
-                        await this.PushBlocksToRepository(context);
+                var canProcessNextBlock = context.CanProcessNextBlock();
+                if (canProcessNextBlock == false)
+                    if (!context.PendingBlockPairsToStore.Any())
```

Comment:
I dont understand this condition 
```
if (canProcessNextBlock == false)
    if (!context.PendingBlockPairsToStore.Any())
        break;
```

if `PendingBlockPairsToStore` has items in it we will not exit the loop and continue to drain `PendingStorage` however in that case all the blocks will return canProcessNextBlock = false right? so there is no need to continue the loop. (please confirm my logic)

I would say: 
```
if(canProcessNextBlock == false)
  break;
```

Reply:
Yes I agree with you, that extra check is not necessary, we need to exit the loop if we cant process the next block.